doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+
+* ``void``
+
+  + ``Void`` (this will return the singleton instance ``nothing::Void``)
+
+* ``T``, where ``T`` is one of the primitive types:
+  ``char``, ``int``, ``long``, ``short``, ``float``, ``double``, ``complex``, ``enum``
+  or any of their `typedef` equivalents
+
+  + ``T``, where ``T`` is an equivalent Julia Bits Type (per the table above)
+  + if ``T`` is an ``enum``, the argument type should be equivalent to ``Cint`` or ``Cuint``
+  + argument value will be copied (returned by-value)
+
+* ``struct T`` (including typedef to a struct)
+
+  + ``T``, where ``T`` is a Julia Leaf Type
+  + argument value will be copied (returned by-value)
+
+* ``void*``
+
+  + depends on how this parameter is used, first translate this to the intended pointer type,
+    then determine the Julia equivalent using the remaining rules in this list
+  + this argument may be declared as ``Ptr{Void}``, if it really is just an unknown pointer
+
+* ``jl_value_t*``
+
+  + ``Any``
+  + argument value must be a valid Julia object
+
+* ``jl_value_t**``
+
+  + ``Ref{Any}``
+  + argument value must be a valid Julia object (or ``C_NULL``)
+
+* ``T*``
+
+  + If the memory is already owned by Julia, or is an `isbits` type, and is known to be non-null:
+
+    + ``Ref{T}``, where ``T`` is the Julia type corresponding to ``T``
+    + a return type of ``Ref{Any}`` is invalid, it should either be ``Any``
+      (corresponding to ``jl_value_t*``) or ``Ptr{Any}`` (corresponding to ``Ptr{Any}``)
+    + currently partially unsupported by cfunction due to #2818
+    + C **MUST NOT** modify the memory returned via ``Ref{T}`` if ``T`` is an ``isbits`` type
+
+  + If the memory is owned by C:
+
+    + ``Ptr{T}``, where ``T`` is the Julia type corresponding to ``T``
+
+* ``(T*)(...)`` (e.g. a pointer to a function)
+
+  + ``Ptr{Void}`` (you may need to use ``cfunction`` explicitly to create this pointer)
 
 Passing Pointers for Modifying Inputs
--------------------------------------
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Because C doesn't support multiple return values, often C functions will take
-pointers to data that the function will modify. To accomplish this within a
-``ccall`` you need to encapsulate the value inside an array of the appropriate
-type. When you pass the array as an argument with a ``Ptr`` type, julia will
-automatically pass a C pointer to the encapsulated data::
+Because C doesn't support multiple return values,
+often C functions will take pointers to data that the function will modify.
+To accomplish this within a ``ccall``,
+you need to first encapsulate the value inside an ``Ref{T}`` of the appropriate type.
+When you pass this ``Ref`` object as an argument,
+julia will automatically pass a C pointer to the encapsulated data::
 
-    width = Cint[0]
-    range = Cfloat[0]
-    ccall(:foo, Void, (Ptr{Cint}, Ptr{Cfloat}), width, range)
+    width = Ref{Cint}(0)
+    range = Ref{Cfloat}(0)
+    ccall(:foo, Void, (Ref{Cint}, Ref{Cfloat}), width, range)
```

Comment:
the explicit typecasts and construction of the `Ref{Cint}` objects are a bit verbose and unnecessary in that case, but yes, it will work